### PR TITLE
provide crypto safe rnglib.

### PR DESCRIPTION
**Pull request by qizhangz**
Fix: #253

Signed-off-by: Qi Zhang <qi1.zhang@intel.com>

### CODE REVIEW

File: os_stub/rnglib_std/rng.c [hunk 1]
```diff
@@ -20,17 +28,44 @@
...
+	int fd;
+
+	fd = open("/dev/urandom", O_RDONLY);
+	if (fd < 0) {
+		printf("cannot open /dev/urandom\n");
+		return FALSE;
+	}
+	if (read(fd, rand_data, sizeof(*rand_data)) != sizeof(*rand_data)) {
+		printf("Cannot read /dev/urandom\n");
+		return FALSE;
```
Comment (jyao1) [inline os_stub/rnglib_std/rng.c:65]: Need close(fd); here.

File: os_stub/rnglib_std/rng.c [hunk 2]
```diff
@@ -20,17 +28,44 @@
...
+		PROV_RSA_FULL,
+		0)) {
+		return FALSE;
+	}
+
+	if(!CryptGenRandom(
+		hCryptProv,
+		sizeof(*rand_data),
+		(BYTE*)rand_data)) {
+		return FALSE;
```
Comment (jyao1) [inline os_stub/rnglib_std/rng.c:47]: Need CryptReleaseContext() here.

Comment (qizhangz) [reply os_stub/rnglib_std/rng.c:49]: done

File: os_stub/rnglib_std/rng.c (hunk 1 above)
Comment (qizhangz) [reply os_stub/rnglib_std/rng.c:68]: done

**Comment by jyao1** [comment]
Looks good

**Comment by jyao1** [comment]
11/29/2021 - had better to provide separate files/libs - std version, windows version, linux version.

File: os_stub/rnglib_std/rng.c [hunk 3]
```diff
@@ -20,17 +28,47 @@
...
+	if(!CryptAcquireContext(
+		&hCryptProv,
+		NULL,
+		NULL,
+		PROV_RSA_FULL,
+		0)) {
+		return FALSE;
+	}
+
+	if(!CryptGenRandom(
```
Comment (rkongintel) [inline os_stub/rnglib_std/rng.c:43]: It looks like this API is deprecated. https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-cryptgenrandom. Perhaps we should use the Crypto Next Gen APIs?

**Comment by qizhangz** [comment]
> 11/29/2021 - had better to provide separate files/libs - std version, windows version, linux version.

done

File: os_stub/rnglib_std/rng.c (hunk 3 above)
Comment (qizhangz) [reply os_stub/rnglib_std/rng.c]: done

File: os_stub/rnglib_std/rng_linux.c [hunk 4]
```diff
@@ -1,36 +1,41 @@
...
+#include <base.h>
+#include <stdlib.h>
+#include "stdio.h"
+#include <unistd.h>
+#include <fcntl.h>
+
+/**
+  Generates a 64-bit random number.
+
+  if rand is NULL, then ASSERT().
```
Comment (steven-bellock) [inline os_stub/rnglib_std/rng_linux.c:16]: Please include NULL check.

File: os_stub/rnglib_std/rng_std.c [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    License: BSD 3-Clause License. For full text see link: https://github.com/DMTF/libspdm/blob/main/LICENSE.md
+**/
+
+#include <base.h>
+#include <stdlib.h>
+
+/**
+  Generates a 64-bit random number.
+
+  if rand is NULL, then ASSERT().
```
Comment (steven-bellock) [inline os_stub/rnglib_std/rng_std.c:13]: Please include NULL check.

File: os_stub/rnglib_std/rng_win.c [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+#include <windows.h>
+#include <bcrypt.h>
+#include <stdio.h>
+
+#pragma comment(lib, "Bcrypt")
+
+/**
+  Generates a 64-bit random number.
+
+  if rand is NULL, then ASSERT().
```
Comment (steven-bellock) [inline os_stub/rnglib_std/rng_win.c:18]: Please include NULL check.

**Review by steven-bellock — CHANGES_REQUESTED** [review]
@jyao1 Should the directory be reverted back to os_stub/rnglib/ since we have both secure and insecure RNG implementations in the same directory?

**Comment by jyao1** [comment]
> reverted back to os_stub/rnglib/

Agree.

File: os_stub/rnglib_std/rng_linux.c (hunk 4 above)
Comment (qizhangz) [reply os_stub/rnglib_std/rng_linux.c]: done

File: os_stub/rnglib_std/rng_std.c (hunk 5 above)
Comment (qizhangz) [reply os_stub/rnglib_std/rng_std.c]: done

File: os_stub/rnglib_std/rng_win.c (hunk 6 above)
Comment (qizhangz) [reply os_stub/rnglib_std/rng_win.c]: done

**Review by steven-bellock — APPROVED** [review]

**Pull request merged**